### ISSUE TITLE
macOS printing ignores printer tray

### ISSUE THREAD

**Issue by tresf** [AdoptOpenJDK/openjdk-support#23]
**Platform:**
OS: macOS 10.14.6 (Mojave), macOS 10.15.1 Beta (Catalina)
JDK: AdoptOpenJDK 11.0.4

**Architecture:**
x86_64

**Hardware:**
* EPSON WF-4740 Series
* Dymo TwinTurbo

**Steps:**

1. Using macOS, set a default printer to a multi-tray printer
2. Insert paper that's identifiable into each tray. (e.g. paper in tray one can be marked `1` with a marker)
3. Run the below code.

**Expected results:**
* Demo prints from last tray.

**Actual results:**
* Demo prints from first tray.

```java
import javax.print.PrintService;
import javax.print.PrintServiceLookup;
import javax.print.attribute.HashPrintRequestAttributeSet;
import javax.print.attribute.PrintRequestAttributeSet;
import javax.print.attribute.standard.Media;
import javax.print.attribute.standard.MediaTray;
import java.awt.*;
import java.awt.print.PageFormat;
import java.awt.print.Printable;
import java.awt.print.PrinterException;
import java.awt.print.PrinterJob;
import java.util.ArrayList;
import java.util.List;

public class PrintLastTray implements Printable {

    private MediaTray lastTray = null;

    public static void main(String ... args) throws PrinterException {
        PrintLastTray lastTrayDemo = new PrintLastTray();

        // Use default printer
        PrintService printer = PrintServiceLookup.lookupDefaultPrintService();

        List<MediaTray> availableTrays = new ArrayList<>();
        Media[] supportedMedia = (Media[])printer.getSupportedAttributeValues(Media.class, null, null);
        for(Media media : supportedMedia) {
            if (media instanceof MediaTray) {
                availableTrays.add((MediaTray)media);
            }
        }

        if (!availableTrays.isEmpty()) {
            PrintRequestAttributeSet attributes = new HashPrintRequestAttributeSet();

            // Use last tray
            lastTrayDemo.lastTray = availableTrays.get(availableTrays.size() - 1);
            System.out.println("Using Tray: " + lastTrayDemo.lastTray);
            attributes.add(availableTrays.get(availableTrays.size() - 1));

            PrinterJob job = PrinterJob.getPrinterJob();
            job.setPrintable(lastTrayDemo);
            job.setPrintService(printer);

            job.print(attributes);
        } else {
            System.err.println("No trays to print to");
        }
    }

    @Override
    public int print(Graphics graphics, PageFormat pf, int pageNumber) {
        Graphics2D g2d = (Graphics2D)graphics;
        g2d.translate(pf.getImageableX(), pf.getImageableY());
        graphics.drawString("This page should print from tray: " + lastTray, 100, 100);

        if (pageNumber == 0) {
            return PAGE_EXISTS;
        } else {
            return NO_SUCH_PAGE;
        }
    }

}
```

**Comment by tresf** [comment]
This issue has been submitted to a private bug tracker (paid OpenJDK support). I will continue to provide updates here as they arrive.

**Comment by karianna** [comment]
You'll also be pleased to hear that it looks like Apple is going to be rejoining OpenJDK, so we may see a little more focus on that port in general.

**Comment by tresf** [comment]
> You'll also be pleased to hear that it looks like Apple is going to be rejoining OpenJDK, so we may see a little more focus on that port in general.

Great news indeed.  I'll ping you on [Slack](https://adoptopenjdk.net/slack.html) to find out more.

**Comment by tresf** [comment]
Some feedback from our 3rd party vendor:

> While we instrument the code with the testcase, we see that though lpr is a supported command for printing in Mac, JDK doesn't make use of it. Unlike linux, JDK in Mac make use of pure native objective C code along with COCOA APIs for the print operation. Our efforts were to identify the `setAttributes` call in the `RasterPrinter` job class and we see that, `setAttributes()` call has been overridden in both linux and MacOS. While the linux has added support for this new attribute now under the following bug.
> 
> [JDK-6357887: selected printertray is ignored under linux](https://bugs.openjdk.java.net/browse/JDK-6357887)
> 
>
> Mac OS overridden `setAttribute()` still doesn't support it. This is the prime reason for why the issue is still present in the Mac. Though the attribute `HashMap` has the `MediaTray` property set into it, it is not communicated to the native code and hence, the setting doesn't apply while printing.
>

A continuing investigation is occuring.

**Comment by tresf** [comment]
The vendor has provided a patched JDK and in my tests it is passing the original unit test.  Next steps, quoting the vendor:

> [We] still need to refactor the code and get it formatted and take it to community before this turns official.

So the upstream bug report should soon be updated to reflect this.

**Comment by tresf** [comment]
The pull request is now available at openjdk: https://github.com/openjdk/jdk/pull/509.  Once merged, it will require a backport to land on JDK11.

**Comment by tresf** [comment]
Merged via https://github.com/openjdk/jdk/commit/3ee0380e99efafbfbb5b75f0c7664095a568fda4.  Next is getting it backported to OpenJDK11.  How does that process work?

**Comment by karianna** [comment]
The original author typically opens up a backport issue in JBS and then emails the jdk-updates list to sanity check if the lead (Andrew Haley) will accept the patch.

**Comment by tresf** [comment]
Backport process [submitted, approved](https://mail.openjdk.java.net/pipermail/jdk-updates-dev/2020-November/004100.html), should be merged to 11u soon.

**Comment by tresf** [comment]
Merged into [OpenJDK11](https://hg.openjdk.java.net/jdk-updates/jdk11u-dev/rev/ba9363426195), closing.
